amqp/src/main/scala/akka/stream/alpakka/amqp/AmqpConnectionProvider.scala

```diff
@@ -0,0 +1,216 @@
+    handshakeTimeout.foreach(factory.setHandshakeTimeout)
+    shutdownTimeout.foreach(factory.setShutdownTimeout)
+    networkRecoveryInterval.foreach(factory.setNetworkRecoveryInterval)
+    automaticRecoveryEnabled.foreach(factory.setAutomaticRecoveryEnabled)
+    topologyRecoveryEnabled.foreach(factory.setTopologyRecoveryEnabled)
+    exceptionHandler.foreach(factory.setExceptionHandler)
+
+    factory.newConnection(hostAndPortList.map(hp => new Address(hp._1, hp._2)).asJava)
+  }
+
+}
+
+object AmqpDetailsConnectionProvider {
+
+  def apply(host: String, port: Int): AmqpDetailsConnectionProvider =
+    AmqpDetailsConnectionProvider(List((host, port)))
+
+  /**
+   * Java API
+   */
+  def create(host: String, port: Int): AmqpDetailsConnectionProvider =
+    AmqpDetailsConnectionProvider(host, port)
+}
+
+object AmqpCredentials {
+
+  /**
+   * Java API
+   */
+  def create(username: String, password: String): AmqpCredentials =
+    AmqpCredentials(username, password)
+}
+
+final case class AmqpCredentials(username: String, password: String) {
+  override def toString = s"Credentials($username, ********)"
+}
+
+object AmqpCachedConnectionProvider {
+
+  /**
+   * Java API
+   */
+  def create(provider: AmqpConnectionProvider): AmqpCachedConnectionProvider =
+    AmqpCachedConnectionProvider(provider: AmqpConnectionProvider)
+
+  /**
+   * Java API
+   */
+  def create(provider: AmqpConnectionProvider, automaticRelease: Boolean): AmqpCachedConnectionProvider =
+    AmqpCachedConnectionProvider(provider: AmqpConnectionProvider, automaticRelease)
+
+  private sealed trait State
+  private case object Empty extends State
+  private case object Connecting extends State
+  private final case class Connected(connection: Connection, clients: Int) extends State
+  private case object Closing extends State
+}
+
+final case class AmqpCachedConnectionProvider(provider: AmqpConnectionProvider, automaticRelease: Boolean = true)
+    extends AmqpConnectionProvider {
+
+  import akka.stream.alpakka.amqp.AmqpCachedConnectionProvider._
+  private val state = new AtomicReference[State](Empty)
+
+  @tailrec
+  override def get: Connection = state.get match {
+    case Empty =>
+      if (state.compareAndSet(Empty, Connecting)) {
+        val connection = provider.get
+        state.compareAndSet(Connecting, Connected(connection, 1))
+        connection
+      } else get
+    case c @ Connected(connection, clients) =>
+      if (state.compareAndSet(c, Connected(connection, clients + 1))) connection
+      else get
+    case _ => get
+  }
+
+  @tailrec
+  override def release(connection: Connection): Unit = state.get match {
+    case c @ Connected(cachedConnection, clients) =>
+      if (cachedConnection != connection)
+        throw new IllegalArgumentException("Can't release a connection that's not owned by this provider")
+
+      if (state.compareAndSet(c, Closing)) {
+        if (clients == 0 || !automaticRelease) {
```

Comment:
What thread will see it get 0 if we are the last client and only subtract in the other branch? It should also not do the Closing swap unless it wants to close, which should only be if clients - 1 becomes 0 and automaticRelease, this you can check before doing the cas.

Reply:
I've updated this to be 1 instead of 0. So when there is only 1 client left, the connection is closed.